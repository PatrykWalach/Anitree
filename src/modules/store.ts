import * as modules from '.'
import { Store, plugin } from 'vuex-composition-api'

export const store =
  process.env.NODE_ENV === 'production'
    ? undefined
    : new Store({
        plugins: [plugin(Object.values(modules))],
      })
