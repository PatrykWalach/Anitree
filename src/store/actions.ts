export { titleActions as title } from './modules/title'

export { editActions as edit } from './modules/edit'
export { filterActions as filter } from './modules/filter'
export { shareActions as share } from './modules/share'
export { settingsActions as settings } from './modules/settings'
export { timelineActions as timeline } from './modules/timeline'
export { changesActions as changes } from './modules/changes'
