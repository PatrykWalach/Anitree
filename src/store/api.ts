import gql from 'graphql-tag'
import { FetchVariables } from './models'

import apollo from '../apollo'
import { AMedia } from '@/types'

export const query = gql`
  query($search: String, $idIn: [Int], $page: Int = 0) {
    Page(page: $page, perPage: 50) {
      media(search: $search, id_in: $idIn) {
        id
        bannerImage
        coverImage {
          extraLarge
          large
          medium
          color
        }
        format
        status
        tags {
          name
          id
          rank
        }
        siteUrl
        type
        studios(isMain: true) {
          nodes {
            name
            id
          }
        }
        startDate {
          year
          month
          day
        }
        endDate {
          year
          month
          day
        }
        title {
          romaji
          english
          native
        }
        description
        mediaListEntry {
          status
        }
        relations {
          edges {
            node {
              id
              type
            }
            id
            relationType
          }
        }
      }
    }
  }
`

export function fetchMediaApollo(variables: FetchVariables): Promise<AMedia[]> {
  return apollo
    .query<{ Page: { media: AMedia[] } }>({
      query,
      variables
    })
    .then(({ data }) => data.Page.media)
}
