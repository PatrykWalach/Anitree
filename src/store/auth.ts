import VueFunctionApi, { ref, Ref, computed } from 'vue-function-api'
import Vue from 'vue'
Vue.use(VueFunctionApi)

import { Token } from '@/types'
const stored = localStorage.getItem('AUTH')

const auth: Ref<Token | null> = ref((stored && JSON.parse(stored)) || null)

const token = computed(() => {
  if (
    auth.value
    // && Date.now() < parseInt(auth.value.expires_in)
  ) {
    return auth.value.access_token
  }
  return null
})

const CHANGE_TOKEN = async (payload: Token | null) => {
  if (payload) {
    const newAuth: Token = {
      ...payload,
      /* eslint-disable-next-line @typescript-eslint/camelcase */
      expires_in: (Date.now() + parseInt(payload.expires_in)).toString()
    }
    localStorage.setItem('AUTH', JSON.stringify(newAuth))
    auth.value = newAuth
  } else {
    auth.value = payload
  }
}

export default function useAuth() {
  return { auth, CHANGE_TOKEN, token, stored }
}
