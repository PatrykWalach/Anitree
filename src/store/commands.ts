import CompositionApi, { Ref, ref, watch } from '@vue/composition-api'
import {
  DeleteCommand,
  DeleteCommandConstructor,
} from './commands/DeleteCommand'
import { DeleteVariables, SaveVariables } from '@/graphql/schema/listEntry'
import { SaveCommand, SaveCommandConstructor } from './commands/SaveCommand'
import { Form } from '@/types'
import { FuzzyDate } from '@/graphql/schema/media'
import { MediaList } from '@/graphql/schema/mediaListCollection'
import Vue from 'vue'
import { useSettings } from './settings'

Vue.use(CompositionApi)

export interface Command {
  saveState?(): void | Promise<void>
  undo?(): void | Promise<void>
  execute(): void | Promise<void>
}

export interface ListCommand extends Command {
  backup: Form | null
  variables: SaveVariables | (DeleteVariables & Pick<SaveVariables, 'mediaId'>)
  done: boolean
  _class: 'SaveCommand' | 'DeleteCommand'
}

export const isDeleteCommand = (
  e: (SaveCommandConstructor | DeleteCommandConstructor) &
    Pick<ListCommand, '_class'>,
): e is DeleteCommandConstructor & Pick<DeleteCommand, '_class'> =>
  e._class === 'DeleteCommand'

export const isSaveCommand = (
  e: (SaveCommandConstructor | DeleteCommandConstructor) &
    Pick<ListCommand, '_class'>,
): e is SaveCommandConstructor & Pick<SaveCommand, '_class'> =>
  e._class === 'SaveCommand'

const stored = localStorage.getItem('CHANGES')

const history: Ref<ListCommand[]> = ref(
  (stored &&
    JSON.parse(stored, (key, value) => {
      if (value instanceof Object) {
        if (isSaveCommand(value)) {
          return new SaveCommand(value)
        }
        if (isDeleteCommand(value)) {
          return new DeleteCommand(value)
        }
      }
      return value
    })) ||
    [],
)

watch(() => {
  if (useSettings().cacheChanges.value) {
    localStorage.setItem('CHANGES', JSON.stringify(history.value))
  }
})
;(async () => {
  for (const command of history.value.filter(({ done }) => !done)) {
    await command.execute()
  }
})()

const dispatch = async (command: ListCommand) => {
  history.value.push(command)
  await command.execute()
  return command
}

export const useCommands = () => {
  return {
    dispatch,
    history,
  }
}

export const mediaListToForm = (
  mediaListEntry: MediaList | null,
  advancedScoring: string[],
): Form => {
  if (mediaListEntry) {
    const advancedScores = advancedScoring
      .map(key => mediaListEntry.advancedScores[key])
      .map(value => value || 0)

    delete mediaListEntry.__typename
    delete mediaListEntry.startedAt.__typename
    delete mediaListEntry.completedAt.__typename
    delete mediaListEntry.private
    delete mediaListEntry.id
    delete mediaListEntry.mediaId

    return {
      ...mediaListEntry,
      advancedScores,
    }
  }

  const advancedScores = advancedScoring.map(() => 0)
  return {
    advancedScores,
    completedAt: { day: null, month: null, year: null },
    notes: '',
    progress: 0,
    progressVolumes: 0,
    repeat: 0,
    score: 0,
    startedAt: { day: null, month: null, year: null },
    status: null,
  }
}

export const isDate = (e: any): e is FuzzyDate => e.__typename === 'FuzzyDate'
