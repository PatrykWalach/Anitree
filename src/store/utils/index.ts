export * from './useCommands'
export * from './useTitle'
export * from './useEdit'
