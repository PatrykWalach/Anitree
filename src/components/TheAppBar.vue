<template>
  <v-app-bar
    app
    elevate-on-scroll
    :flat="routeTitle && !$vuetify.breakpoint.xsOnly"
  >
    <v-tooltip v-if="!routeHome && !routeSearch" bottom>
      <template v-slot:activator="{ on, attrs }">
        <v-btn icon v-bind="attrs" v-on="on" @click.stop="$router.back()"
          ><v-icon>arrow_back</v-icon></v-btn
        >
      </template>
      <span>Back</span>
    </v-tooltip>

    <template v-if="extension" v-slot:extension>
      <TheAppBarExtensionTabs v-if="tabs" />
      <TheAppBarExtensionChips v-else />
    </template>

    <template v-if="!routeSearch">
      <v-toolbar-title class="text-capitalize">
        <span v-if="routeTitle" :title="title">
          {{ title }}
        </span>
        <template v-else-if="!routeHome">
          {{ $route.name }}
        </template>
        <template v-else> Anitree <sup class="overline">alpha</sup> </template>
      </v-toolbar-title>
    </template>

    <TheAppBarSearch v-if="routeSearch" />

    <template v-if="!$vuetify.breakpoint.xsOnly">
      <v-spacer v-if="filteredActions.length"></v-spacer>
      <v-tooltip
        bottom
        :key="title"
        v-for="{ icon, bind, title, on: actionOn } in mainActions"
      >
        <template v-slot:activator="{ attrs, on }">
          <v-btn
            v-bind="{ ...bind, ...attrs }"
            icon
            v-on="{ ...on, ...actionOn }"
          >
            <v-icon>{{ icon }}</v-icon>
          </v-btn>
        </template>
        <span>{{ title }}</span>
      </v-tooltip>
      <TheAppBarMenu :media="media" bottom :actions="moreActions" />
    </template>
  </v-app-bar>
</template>
<script lang="ts">
import { SetupContext, computed, createComponent } from '@vue/composition-api'
import { Media } from '../graphql/schema/media'
import TheAppBarSearch from './TheAppBarSearch.vue'
const TheAppBarExtensionChips = () =>
  import(
    /* webpackChunkName: "TheAppBarExtensionChips" */ /* webpackPrefetch: true */ './TheAppBarExtensionChips.vue'
  )
const TheAppBarExtensionTabs = () =>
  import(
    /* webpackChunkName: "TheAppBarExtensionTabs" */ /* webpackPrefetch: true */ './TheAppBarExtensionTabs.vue'
  )
const TheAppBarMenu = () =>
  import(
    /* webpackChunkName: "TheAppBarMenu" */ /* webpackPrefetch: true */ './TheAppBarMenu.vue'
  )

import { useAppBarActions } from './TheBottomAppBar.vue'
import { useFab } from './TheFab.vue'

import { useNavigation } from './TheDrawer.vue'
import { useTitle } from '../store'

export const useRoutes = (root: SetupContext['root']) => {
  const routeTitle = computed(() => {
    const { name } = root.$route
    return name === 'media-about' || name === 'media-timeline'
  })
  const routeHome = computed(() => root.$route.name === 'home')
  const routeSearch = computed(() => root.$route.name === 'search')

  const { main } = useNavigation()

  const routeMain = computed(() => {
    const { name } = root.$route

    return !!main.find(({ bind }) => bind.to.name === name)
  })

  return {
    routeHome,
    routeMain,
    routeSearch,
    routeTitle,
  }
}
export interface Props {
  media: Media | null
}

export default createComponent<Readonly<Props>>({
  components: {
    TheAppBarExtensionChips,
    TheAppBarExtensionTabs,
    TheAppBarMenu,
    TheAppBarSearch,
  },
  props: {
    media: { default: null, required: true, type: null },
  },
  setup(props, { root }) {
    const { routeSearch, routeHome, routeTitle } = useRoutes(root)

    const { getTitle } = useTitle()

    const title = computed(() => {
      const media = props.media
      return getTitle.value(media && media.title)
    })

    const tabs = computed(
      () => routeTitle.value && root.$vuetify.breakpoint.xsOnly,
    )

    const extension = computed(
      () =>
        tabs.value ||
        (routeSearch.value && !!Object.keys(root.$route.query).length),
    )

    return {
      ...useFab(root),
      ...useAppBarActions(props, root),
      extension,
      routeHome,
      routeSearch,
      routeTitle,
      tabs,
      title,
    }
  },
})
</script>
