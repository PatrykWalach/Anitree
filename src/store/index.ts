import Vue from 'vue'
import Vuex from 'vuex'

// import { getModule } from 'vuex-module-decorators'
Vue.use(Vuex)

import { ModuleMedia } from './modules/media'
import { ModuleTheme } from './modules/theme'
import { ModuleTitle } from './modules/title'
import { ModuleAuth } from './modules/auth'

export interface RootState {
  media: ModuleMedia
  theme: ModuleTheme
  title: ModuleTitle
  auth: ModuleAuth
}

export default new Vuex.Store<RootState>({})
