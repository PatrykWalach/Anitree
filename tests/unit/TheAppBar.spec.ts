import { createVariations, matchSnapshot } from './utils'
import TheAppBar from '@/components/TheAppBar.vue'
import { useMockedStore } from './mocks/store'


describe('TheAppBar', () => {
  createVariations({
    mocks: {
      $route: () => [
        ...createVariations({
          name: () => ['home', 'settings', 'search'],
          query: {},
        }),
        {
          name: 'search',
          query: { search: 'test' },
        },
      ],
    },
  }).forEach(settings =>
    matchSnapshot(TheAppBar, {
      propsData: { media: null },
      ...settings,
      ...useMockedStore(),
    }),
  )

  // matchSnapshot(
  //   TheAppBar,
  //   {
  //     mocks: {
  //       $route: {
  //         name: 'media-about',
  //         query: {},
  //       },
  //       $vuetify: {
  //         breakpoint: {
  //           xsOnly: false,
  //         },
  //       },
  //     },
  //     propsData: { media: mockedMedia },
  //     ...useMockedStore(),
  //   },
  //   'hides media tabs',
  // )

  // matchSnapshot(
  //   TheAppBar,
  //   {
  //     mocks: {
  //       $route: {
  //         name: 'media-about',
  //         query: {},
  //       },
  //       $vuetify: {
  //         breakpoint: {
  //           xsOnly: true,
  //         },
  //       },
  //     },
  //     propsData: { media: mockedMedia },
  //     ...useMockedStore(),
  //   },
  //   'shows media tabs',
  // )
})
