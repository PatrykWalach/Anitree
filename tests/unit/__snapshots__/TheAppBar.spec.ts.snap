// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheAppBar matches snapshot 1`] = `
<v-app-bar-stub
  app="true"
  elevateonscroll="true"
  extensionheight="48"
  src=""
  tag="header"
  tile="true"
  value="true"
>
  <!---->
    
  <v-toolbar-title-stub
    class="text-capitalize"
  >
     Anitree 
    <sup
      class="overline"
    >
      alpha
    </sup>
  </v-toolbar-title-stub>
   
  <!---->
   
  <!---->
    
  <!---->
</v-app-bar-stub>
`;

exports[`TheAppBar matches snapshot 2`] = `
<v-app-bar-stub
  app="true"
  elevateonscroll="true"
  extensionheight="48"
  src=""
  tag="header"
  tile="true"
  value="true"
>
  <v-tooltip-stub
    bottom="true"
    closedelay="0"
    contentclass=""
    fixed="true"
    maxwidth="auto"
    nudgebottom="0"
    nudgeleft="0"
    nudgeright="0"
    nudgetop="0"
    nudgewidth="0"
    opendelay="0"
    openonhover="true"
    tag="span"
  >
     
    <span>
      Back
    </span>
  </v-tooltip-stub>
    
  <v-toolbar-title-stub
    class="text-capitalize"
  >
    
        settings
      
  </v-toolbar-title-stub>
   
  <!---->
   
  <!---->
    
  <!---->
</v-app-bar-stub>
`;

exports[`TheAppBar matches snapshot 3`] = `
<v-app-bar-stub
  app="true"
  elevateonscroll="true"
  extensionheight="48"
  src=""
  tag="header"
  tile="true"
  value="true"
>
  <!---->
    
  <!---->
   
  <theappbarsearch-stub />
   
  <!---->
    
  <!---->
</v-app-bar-stub>
`;

exports[`TheAppBar matches snapshot 4`] = `
<v-app-bar-stub
  app="true"
  elevateonscroll="true"
  extensionheight="48"
  src=""
  tag="header"
  tile="true"
  value="true"
>
  <!---->
    
  <!---->
   
  <theappbarsearch-stub />
   
  <!---->
    
  <!---->
</v-app-bar-stub>
`;
